[GTK][WPE] Add API to check if a response policy decision is for the main resource

carlosgcampos opened the pull request:
#### e35d40906d10a1494e117ab696ab323191803f7c
<pre>
[GTK][WPE] Add API to check if a response policy decision is for the main resource
<a href="https://bugs.webkit.org/show_bug.cgi?id=247059">https://bugs.webkit.org/show_bug.cgi?id=247059</a>

Reviewed by Adrian Perez de Castro and Michael Catanzaro.

Apps are currently checking if request uri is the same as web view main
resource uri which might not be accurate.

* Source/WebKit/UIProcess/API/glib/WebKitResponsePolicyDecision.cpp:
(webkit_response_policy_decision_is_main_frame_main_resource):
* Source/WebKit/UIProcess/API/glib/WebKitResponsePolicyDecision.h.in:
* Tools/MiniBrowser/gtk/BrowserTab.c:
(decidePolicy):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitPolicyClient.cpp:
(testResponsePolicy):

Canonical link: <a href="https://commits.webkit.org/256058@main">https://commits.webkit.org/256058@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ffd8f2c74ed180e4eaf752e31e8a19c662eb2db

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94332 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3509 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27261 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103999 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164273 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98330 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3554 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31716 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86668 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100001 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100002 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2548 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80726 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29595 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84468 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/84090 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72486 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38111 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17917 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35988 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19189 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4186 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39875 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41792 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41824 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38418 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitPolicyClient.cpp:
```diff
@@ -222,6 +222,7 @@ static void testResponsePolicy(PolicyClientTest* test, gconstpointer)
     ASSERT_CMP_CSTRING(webkit_uri_response_get_uri(response), ==, kServer->getURIForPath("/"));
     g_assert_cmpint(webkit_web_view_can_show_mime_type(test->m_webView, webkit_uri_response_get_mime_type(response)), ==,
         webkit_response_policy_decision_is_mime_type_supported(decision));
+    g_assert_true(webkit_response_policy_decision_is_main_frame_main_resource(decision));
```
Would be nice to test that it can return FALSE as well.

carlosgcampos replied:
I agree, I'll add a new test in a follow up, becase it's not easy with current tests that always expect a single policy decision.

carlosgcampos committed:
[GTK][WPE] Add API to check if a response policy decision is for the main resource
https://bugs.webkit.org/show_bug.cgi?id=247059

Reviewed by Adrian Perez de Castro and Michael Catanzaro.

Apps are currently checking if request uri is the same as web view main
resource uri which might not be accurate.

* Source/WebKit/UIProcess/API/glib/WebKitResponsePolicyDecision.cpp:
(webkit_response_policy_decision_is_main_frame_main_resource):
* Source/WebKit/UIProcess/API/glib/WebKitResponsePolicyDecision.h.in:
* Tools/MiniBrowser/gtk/BrowserTab.c:
(decidePolicy):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitPolicyClient.cpp:
(testResponsePolicy):

Canonical link: https://commits.webkit.org/256058@main
